proposals/NNNN-opt-in-reflection-metadata.md

```diff
@@ -0,0 +1,181 @@
+*   Authors: [Max Ovtsin](https://github.com/maxovtsin)
+*   Review Manager: TBD
+*   Status: **Awaiting implementation**
+*   Implementation: [apple/swift#34199](https://github.com/apple/swift/pull/34199)
+
+## Introduction
+
+This proposal seeks to increase the safety, efficiency and privacy of Swift Reflection Metadata by improving the existing mechanism and providing the opportunity to express a requirement on Reflection Metadata in APIs that consume it. 
+  
+Swift-evolution thread: [Discussion thread topic for that proposal](https://forums.swift.org/t/pitch-3-opt-in-reflection-metadata/58852)
+
+
+## Motivation
+
+There are two kinds of Swift metadata emitted be the compiler:
+
+1. Core Metadata (type metadata record, nominal type descriptor, etc).
+2. Reflection metadata (reflection metadata field descriptor).
+
+Core metadata must constantly be emitted and may only be stripped if provenly not used. (This metadata isn't affected by this proposal)
+Reflection metadata contains optional information about declarations' fields - their names and references to their types. This metadata isn't used by the language's runtime features, and the emission may be skipped if such types aren't passed to reflection-consuming APIs.
+
+APIs can use Reflection Metadata differently. Some like `print`, and `dump` will still work with disabled reflection, but the output will be limited. Others, like SwiftUI, rely on it and won't work correctly if the reflection metadata is missing.
+While the former can benefit as well, the main focus of this proposal is on the latter.
+
+A developer can mistakenly turn off Reflection Metadata for a Swift module and won't be warned at compile-time if APIs that consume reflection are used by that module. An app with such a module won't behave as expected at runtime which may be challenging to notice and track down such bugs back to Reflection. For instance, SwiftUI implementation uses reflection metadata from user modules to trigger the re-rendering of the view hierarchy when a state has changed. If for some reason a user module was compiled with metadata generation disabled, changing the state won't trigger that behavior and will cause inconsistency between state and representation which will make such API less safe since it becomes a runtime issue rather than a compile-time one.
+
+On the other hand, excessive Reflection metadata may be preserved in a binary even if not used, because there is currently no way to statically determine its usage. There was an attempt to limit the amount of unused reflection metadata by improving its stripability by the Dead Code Elimination LLVM pass, but in many cases, it’s still preserved in the binary because it’s referenced by Full Type Metadata which prevents Reflection Metadata from stripping.  This unnecessarily increases the binary size and may affect privacy by storing more information about the code's semantics in a binary, which might be used for reverse-engineering.
```

Comment:
I think "privacy" is not the right word here.  We usually use "privacy" to refer to the end-user's personal data, not to trade secrets of the application author.  How about just:

   "This increases binary size and may simplify reverse-engineering."

Reply:
I used the word privacy to emphasize that reverse engineering may help to find some vulnerabilities in an app, which may lead to stealing user data. But I agree that the connection isn't obvious here.